domains/client/domain-executor/src/core_bundle_processor.rs

```diff
@@ -115,11 +121,41 @@ where
         self,
         primary_info: (PBlock::Hash, NumberFor<PBlock>, ForkChoiceStrategy),
     ) -> Result<(), sp_blockchain::Error> {
-        let parent_hash = self.client.info().best_hash;
-        let parent_number = self.client.info().best_number;
+        tracing::debug!(?primary_info, "Processing imported primary block");
 
-        self.process_bundles_at(primary_info, (parent_hash, parent_number))
-            .await?;
+        let (primary_hash, primary_number, fork_choice) = primary_info;
+
+        let ForkAwareBlockImports {
+            initial_parent,
+            primary_imports,
+        } = self
+            .domain_block_processor
+            .fork_aware_block_imports(primary_hash, primary_number)?;
+
+        tracing::trace!(
+            ?initial_parent,
+            ?primary_imports,
+            "Pending primary blocks to process"
+        );
+
+        let mut domain_parent = initial_parent;
+
+        for (i, primary_info) in primary_imports.iter().enumerate() {
+            // Use the origin fork_choice for the target primary block,
+            // the intermediate ones use `Custom(false)`.
+            let fork_choice = if i == primary_imports.len() - 1 {
```

Comment:
Sorry to reopen this, but would be great to avoid the index check. Re: `primary_info == primary_imports.last()` check: can lead to false positives if there are duplicates, if the last element happens to be present early on in the iteration (not sure this applies in this case) 

Other solution would be use `peekable()`, something like:

```
let mut iter = primary_imports.iter().peekable();
while Some(primary_info) = iter.next() {
    let fork_choice = iter.peek().map_or(ForkChoiceStrategy::Custom(false), || fork_choice);
    // ...
}
```

Reply:
There will be no duplicates here, but peekable looks nicer indeed, I always forget about is